# Fuel Management Project

## Overview

This project aims to analyze and optimize fuel usage within a specified system. The primary objectives include data cleaning, preprocessing, exploratory data analysis (EDA), and the development of predictive models to improve fuel management efficiency.

## Contents

1. **Data Cleaning and Preprocessing**: Steps to handle missing values, remove duplicates, and normalize data.
3. **Feature Engineering**: Creation of new features or transformation of existing ones to improve model performance.
5. **Model Evaluation**: Assessment of model performance using appropriate metrics and validation techniques.
6. **Conclusions and Insights**: Summary of findings and recommendations for improving fuel management.


# 1. Data Loading and Overview

In [8]:
#1.1: Importing the essential Libraries
import pandas as pd
import numpy as np
import math
import matplotlib.pyplot as plt
import seaborn as sns

In [9]:
#1.2: Loading the data
df = pd.read_csv("Final Data.csv")
df.head(10)

Number Registration                              Details Product Odometer  \
0    NaN          NaN                                  NaN     NaN      NaN   
1   0001           U1  Jaguar XJL Portfolio V6 D Limousine  Diesel    67691   
2   0012      FL65FBA             Vauxhall CORSA  CDTI Car  Diesel     9329   
3   0013      MJ16XMY   Vauxhall Mokka 4x4 Utility Vehicle  Diesel    29064   
4   0016      BP16MKK                      Nissan Note Car  Diesel    42233   
5   0017      YD16NAL                      Peugeot car Car  Diesel    20475   
6   0018      YD16NAK                  Peugeot 308 Car Car  Diesel   115020   
7   0019      BP16OPD                      Nissan Note Car  Diesel    15559   
8   0020      YH17VCF      Peugeot Expert Combi 9 seat MPV  Diesel    37352   
9   0021      MF68FNR             Renault  Traffic Welfare  Diesel      112   

  Distance   Unit     MPG  
0      NaN    NaN     NaN  
1        0  Miles    0.00  
2     2849  Miles   50.04  
3     5369  Miles   39.29  
4     1052  Miles   56.17  
5     3319  Miles  108.32  
6    36100  Miles   50.79  
7      948  Miles   36.04  
8    14202  Miles   34.75  
9        0  Miles    0.00

In [10]:
# 1.3: Looking for data types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1490 entries, 0 to 1489
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Number        1486 non-null   object 
 1   Registration  1483 non-null   object 
 2   Details       1484 non-null   object 
 3   Product       1484 non-null   object 
 4   Odometer      1484 non-null   object 
 5   Distance      1484 non-null   object 
 6   Unit          1482 non-null   object 
 7   MPG           1482 non-null   float64
dtypes: float64(1), object(7)
memory usage: 93.3+ KB


## Dataset Overview

The dataset contains 1,490 entries with 8 columns related to fuel management. Below is a brief description of the dataset structure:

- **Number**: Identification number for each entry (1486 non-null values, dtype: object)
- **Registration**: Vehicle registration details (1483 non-null values, dtype: object)
- **Details**: Additional details about the entry (1484 non-null values, dtype: object)
- **Product**: Type of fuel product used (1484 non-null values, dtype: object)
- **Odometer**: Odometer reading at the time of fuel entry (1484 non-null values, dtype: object)
- **Distance**: Distance covered (1484 non-null values, dtype: object)
- **Unit**: Unit of measurement for fuel (1482 non-null values, dtype: object)
- **MPG**: Miles per gallon (1482 non-null values, dtype: float64)

The dataset has some missing values across several columns, which will be addressed during the data cleaning and preprocessing phase.

In [11]:
#1.4: Statistical significance of the data
df.describe(include = 'all')

Number Registration                    Details Product Odometer  \
count    1486         1483                       1484    1484     1484   
unique   1485         1452                        301       5     1359   
top      1482      MW16DDL  Peugeot Partner Van Small  Diesel        0   
freq        2            2                        104    1440      114   
mean      NaN          NaN                        NaN     NaN      NaN   
std       NaN          NaN                        NaN     NaN      NaN   
min       NaN          NaN                        NaN     NaN      NaN   
25%       NaN          NaN                        NaN     NaN      NaN   
50%       NaN          NaN                        NaN     NaN      NaN   
75%       NaN          NaN                        NaN     NaN      NaN   
max       NaN          NaN                        NaN     NaN      NaN   

       Distance   Unit          MPG  
count      1484   1482  1482.000000  
unique     1187      4          NaN  
top           0  Miles          NaN  
freq        250   1265          NaN  
mean        NaN    NaN    12.037726  
std         NaN    NaN   246.914654  
min         NaN    NaN -9238.620000  
25%         NaN    NaN     3.167500  
50%         NaN    NaN    19.750000  
75%         NaN    NaN    29.885000  
max         NaN    NaN   824.850000

### Summary Statistics

- **Number**
  - Count: 1486
  - Unique: 1485
  - Most frequent (top): 1482 (2 occurrences)

- **Registration**
  - Count: 1483
  - Unique: 1452
  - Most frequent (top): MW16DDL (2 occurrences)

- **Details**
  - Count: 1484
  - Unique: 301
  - Most frequent (top): Peugeot Partner Van Small (104 occurrences)

- **Product**
  - Count: 1484
  - Unique: 5
  - Most frequent (top): Diesel (1440 occurrences)

- **Odometer**
  - Count: 1484
  - Unique: 1359
  - Most frequent (top): 0 (114 occurrences)

- **Distance**
  - Count: 1484
  - Unique: 1187
  - Most frequent (top): 0 (250 occurrences)

- **Unit**
  - Count: 1482
  - Unique: 4
  - Most frequent (top): Miles (1265 occurrences)

- **MPG**
  - Count: 1482
  - Mean: 12.0377
  - Standard Deviation: 246.9147
  - Minimum: -9238.62
  - 25th Percentile: 3.1675
  - 50th Percentile (Median): 19.75
  - 75th Percentile: 29.885
  - Maximum: 824.85

### Notes

- The dataset has some missing values across several columns, which will be addressed during the data cleaning and preprocessing phase.
- The `MPG` column has some extreme values, which may need further investigation for outliers or data entry errors.

# 2. Data Cleaning

## 2.1: Handling Null Values

In [15]:
#Checking for null values
df.isna().sum()

Details     6
Product     6
Odometer    6
Distance    6
Unit        8
MPG         8
dtype: int64

In [37]:
#Percentage of NULL values
total_values = df.shape[0] * df.shape[1]  #This will return the total values that df has
null_values = df.isnull().sum().sum()     #This will return the total null values from all the df
per_null_values = (null_values*100)/total_values     #This will return the percentage of null values

print("The percentage of the null values in this data is:", per_null_values.round(2))

The percentage of the null values in this data is: 0.45


#### As this datafram has less than 1% null values so dropping these values will be used.

In [40]:
#Dropping the null values
df.dropna(axis = 0, inplace = True)
df.head()

Details Product Odometer Distance   Unit  \
1  Jaguar XJL Portfolio V6 D Limousine  Diesel    67691        0  Miles   
2             Vauxhall CORSA  CDTI Car  Diesel     9329     2849  Miles   
3   Vauxhall Mokka 4x4 Utility Vehicle  Diesel    29064     5369  Miles   
4                      Nissan Note Car  Diesel    42233     1052  Miles   
5                      Peugeot car Car  Diesel    20475     3319  Miles   

      MPG  
1    0.00  
2   50.04  
3   39.29  
4   56.17  
5  108.32

In [41]:
#Check for any null values 
df.isnull().sum()

Details     0
Product     0
Odometer    0
Distance    0
Unit        0
MPG         0
dtype: int64

## 2.2: Handling Duplicate Values

In [47]:
# Checking for duplicates
df.duplicated().sum()

55

In [50]:
#The actual values that are duplicate
duplicate_values = df[df.duplicated()]
duplicate_values

Details Product Odometer Distance   Unit  MPG
29      Volkswagen Maxi Life Caddy  Diesel        0        0  Miles  0.0
32      Volkswagen Maxi Life Caddy  Diesel        0        0  Miles  0.0
35        Volkwagon Caddy Life MPV  Diesel        0        0  Miles  0.0
38       Volkswagon Caddy life MPV  Diesel        0        0  Miles  0.0
40       Volkswagon Caddy Life MPV  Diesel        0        0  Miles  0.0
43                      Custom WAV  Diesel        0        0  Miles  0.0
346             Toyota Hilux 4 X 4  Diesel        0        0  Miles  0.0
347             Toyota Hilux 4 X 4  Diesel        0        0  Miles  0.0
451       John Deere 5090R Tractor  Diesel        0        0  Hours  0.0
453       John Deere 5090R Tractor  Diesel        0        0  Hours  0.0
457       John Deere 5090R Tractor  Diesel        0        0  Hours  0.0
486       Vauxhall Corsa Small Van  Diesel        0        0  Miles  0.0
504       Vauxhall Corsa Small Van  Diesel        0        0  Miles  0.0
524       Vauxhall Corsa Small Van  Diesel        0        0  Miles  0.0
535       Vauxhall Corsa Small Van  Diesel        0        0  Miles  0.0
554       Vauxhall Corsa Small Van  Diesel        0        0  Miles  0.0
565       Vauxhall Corsa Small Van  Diesel        0        0  Miles  0.0
572      Vauxhall  Corsa Small Van  Diesel        0        0  Miles  0.0
581       Vauxhall Corsa Small Van  Diesel        0        0  Miles  0.0
602      Vauxhall  Corsa Small Van  Diesel        0        0  Miles  0.0
609       Vauxhall Corsa Small Van  Diesel        0        0  Miles  0.0
613       Vauxhall Corsa Small Van  Diesel        0        0  Miles  0.0
640       John Deere 5090R Tractor  Diesel        0        0  Hours  0.0
642       John Deere 5090R Tractor  Diesel        0        0  Hours  0.0
690           John Deere Gator ATV  Diesel        0        0  Hours  0.0
1022             Drop Side Transit  Diesel        0        0  Miles  0.0
1106      FORD Transit 3.5t Tipper  Diesel        0        0  Miles  0.0
1107      FORD Transit 3.5t Tipper  Diesel        0        0  Miles  0.0
1108      FORD Transit 3.5t Tipper  Diesel        0        0  Miles  0.0
1110                   3.5t Tipper  Diesel        0        0  Miles  0.0
1111                   3.5t Tipper  Diesel        0        0  Miles  0.0
1112                   3.5t Tipper  Diesel        0        0  Miles  0.0
1113                   3.5t Tipper  Diesel        0        0  Miles  0.0
1115               Crew Cab Tipper  Diesel        0        0  Miles  0.0
1116               Crew Cab Tipper  Diesel        0        0  Miles  0.0
1117               Crew Cab Tipper  Diesel        0        0  Miles  0.0
1118               Crew Cab Tipper  Diesel        0        0  Miles  0.0
1119               Crew Cab Tipper  Diesel        0        0  Miles  0.0
1120               Crew Cab Tipper  Diesel        0        0  Miles  0.0
1121               Crew Cab Tipper  Diesel        0        0  Miles  0.0
1122               Crew Cab Tipper  Diesel        0        0  Miles  0.0
1123               Crew Cab Tipper  Diesel        0        0  Miles  0.0
1124               Crew Cab Tipper  Diesel        0        0  Miles  0.0
1127  Ford Transit Crew Cab Tipper  Diesel        0        0  Miles  0.0
1170           Iveco 7Tonne Tipper  Diesel        0        0  Miles  0.0
1171           Iveco 7Tonne Tipper  Diesel        0        0  Miles  0.0
1172           Iveco 7Tonne Tipper  Diesel        0        0  Miles  0.0
1392             Ford Ranger 4 x 4  Diesel        0        0  Miles  0.0
1393      Vauxhall Corsa Small Van  Diesel        0        0  Miles  0.0
1394      Vauxhall Corsa Small Van  Diesel        0        0  Miles  0.0
1395     Vauxhall  Corsa Small Van  Diesel        0        0  Miles  0.0
1397      Vauxhall Corsa Small Van  Diesel        0        0  Miles  0.0
1398     Vauxhall  Corsa Small Van  Diesel        0        0  Miles  0.0
1480               Peugeot 208 Car  Diesel        0        0  Miles  0.0
1482        

#### As this data is illustrating that most of the values from this dataframe are zero in distance and zero in MPG so removing these values might be harmfull for the model because these might be the only zero's values in the dataframe.

So first let's check the unique values across each column

In [65]:
unique_list = []
def uniques_values(raw_data):
    for columns in raw_data.columns:
        print("These are the unique values from the:", columns, " column")
        print("<--------------------------------------------------------------------------------------------->")
        print(raw_data[columns].unique())
        print("The count for all the unique values is:", raw_data[columns].value_counts())
        
uniques_values(df)

These are the unique values from the: Details  column
<--------------------------------------------------------------------------------------------->
['Jaguar XJL Portfolio V6 D Limousine' 'Vauxhall CORSA  CDTI Car'
 'Vauxhall Mokka 4x4 Utility Vehicle' 'Nissan Note Car' 'Peugeot car Car'
 'Peugeot 308 Car Car' 'Peugeot Expert Combi 9 seat MPV'
 'Renault  Traffic Welfare' 'Volkswagon Caddy Maxi Welfare'
 '  Caddy Maxi Life Wav' '  Caddy Maxi Life Van'
 'Volkswagen  Caddy Maxi Life Van' '  VW Caddy'
 'Volkswagen Caddy 7 seater' 'Volkswagen Maxi Life Caddy'
 'Volkwagon Caddy Life MPV' 'Volkswagon Caddy MPV Van Small'
 'Volkswagen Caddy ' 'Volkswagon Caddy life MPV'
 'Volkswagon Caddy WAV Van Small' 'Volkswagon Caddy Life MPV'
 '  Custom WAV' 'Ford Kuga Zetec 4x4 Utility Vehicle'
 'Peugeot Expert Welfare MPV Welfare' 'Ford Mondeo car Car'
 'Peugeot Expert Tepee MPV' 'Peugeot 208 Car' 'Peugeot 308 Estate Car'
 'Peugeot  7 Seater' 'Ford Fiesta 1.4 TDi Van Car Derived Van'
 'Vauxhall Corsa S

#### So as there are more than 250 values with 0 field so it is good to move with removing the duplicates but left the last one and remove the remaining.

In [66]:
df.drop_duplicates(keep= 'last', inplace=True)

## 2.3: Extracting the Car Brand Name

In [91]:
#Getting the car brand name by uisng split and lambda methods
try:
    df['Car_name'] = df['Details'].apply(lambda x: x.split("")[0])
except:
    print("Error")

Error


In [92]:
df['Car_name'].unique()

array(['Jaguar', 'Vauxhall', 'Nissan', 'Peugeot', 'Renault', 'Volkswagon',
       '', 'Volkswagen', 'Volkwagon', 'Ford', 'Mercedes', 'Fiat',
       'Mitsubishi', 'Toyota', 'John', 'chipper', 'PEUGEOT', 'Valtra',
       'peugeot', 'Kubota', 'Toro', 'Torro', 'JD', 'Timberwolf', 'JCB',
       'Carlton', 'Volvo', 'FORD', 'Renauly', 'TS', 'Iveco', 'vauxha;;',
       'Vauxall', 'Vauvhall', 'Leyland', 'Fuso', 'Daf', 'Dennis',
       'Mereecdes', 'DAF', 'vauxhall', 'Landrover', 'LDV', 'Johnson',
       'Linde', 'MPM'], dtype=object)

In [93]:
df[df['Car_name'] == '']

Details Product Odometer Distance   Unit    MPG Car_name
11      Caddy Maxi Life Wav  Diesel     8103     8066  Miles  42.22         
12      Caddy Maxi Life Van  Diesel     6187     6160  Miles  36.93         
14                 VW Caddy  Diesel     1060     1032  Miles  26.60         
15                 VW Caddy  Diesel      360      307  Miles  15.77         
16                 VW Caddy  Diesel      967      884  Miles  24.48         
...                     ...     ...      ...      ...    ...    ...      ...
1442          Small Sweeper  Diesel    28548     4033  Miles   4.33         
1443          Small Sweeper  Diesel    22090     5901  Miles   4.41         
1444          Small Sweeper  Diesel    33210     6513  Miles   3.83         
1445          Small Sweeper  Diesel    37376    10179  Miles   5.77         
1446          Small Sweeper  Diesel    27310     7305  Miles   4.14         

[239 rows x 7 columns]

# 3. Feature Engineering

In [7]:
#1.5: Removing the unwanted columns but first let's look again the data we have
df.head()

Details Product Odometer Distance   Unit    MPG
0                                  NaN     NaN      NaN      NaN    NaN    NaN
1  Jaguar XJL Portfolio V6 D Limousine  Diesel    67691        0  Miles   0.00
2             Vauxhall CORSA  CDTI Car  Diesel     9329     2849  Miles  50.04
3   Vauxhall Mokka 4x4 Utility Vehicle  Diesel    29064     5369  Miles  39.29
4                      Nissan Note Car  Diesel    42233     1052  Miles  56.17

## 3.1 Unwanted Columns:
1. Number:  because it has no affect on the fuel management it is just a index type
2. Registration: It could be useful for data management but not here

In [12]:
# Dropping the columns
df.drop(['Number', 'Registration'], axis = 1, inplace = True)
df.head()

Details Product Odometer Distance   Unit    MPG
0                                  NaN     NaN      NaN      NaN    NaN    NaN
1  Jaguar XJL Portfolio V6 D Limousine  Diesel    67691        0  Miles   0.00
2             Vauxhall CORSA  CDTI Car  Diesel     9329     2849  Miles  50.04
3   Vauxhall Mokka 4x4 Utility Vehicle  Diesel    29064     5369  Miles  39.29
4                      Nissan Note Car  Diesel    42233     1052  Miles  56.17